src/java.base/share/classes/sun/security/ssl/PreSharedKeyExtension.java

```diff
@@ -369,23 +369,33 @@ public void consume(ConnectionContext context,
                 SSLSessionContextImpl sessionCache = (SSLSessionContextImpl)
                         shc.sslContext.engineGetServerSessionContext();
                 int idIndex = 0;
-                SSLSessionImpl s = null;
 
                 for (PskIdentity requestedId : pskSpec.identities) {
+                    SSLSessionImpl s = null;
                     // If we are keeping state, see if the identity is in the cache
                     if (requestedId.identity.length == SessionId.MAX_LENGTH) {
-                        s = sessionCache.get(requestedId.identity);
-                    }
-                    // See if the identity is a stateless ticket
-                    if (s == null &&
-                            requestedId.identity.length > SessionId.MAX_LENGTH &&
-                            sessionCache.statelessEnabled()) {
+                        synchronized (sessionCache) {
```

Comment:
Did you have a test if there is a performance regression by introducing the synchronization here?

Reply:
Yes, I’ve made a test that calculates total time spent by server to receive <N> connections. Every server handshake is performed in a separate thread
The client starts <T> threads. Every thread sends one initial connection and <R-1> renegotiated connections. So, the total number of connections is  <N>=<T>*<R>

Results for the original and JDK-8241248 code are almost identical:
<T>=10 <R>=100
Original OpenJDK: 1140 ms
JDK-8241248 code: 1090 ms

<T>=50 <R>=100
Original OpenJDK: 1164 ms
JDK-8241248 code: 1108 ms

<T>=60 <R>=100
Original OpenJDK: 1461 ms
JDK-8241248 code: 1579 ms

<T>=70 <R>=100
Original OpenJDK: 2058 ms
JDK-8241248 code: 1999 ms

<T>=80 <R>=100
Original OpenJDK: 2148 ms
JDK-8241248 code: 2125 ms

<T>=90 <R>=100
Original OpenJDK: 2540 ms
JDK-8241248 code: 2514 ms

<T>=90 <R>=100
Original OpenJDK: 3011 ms
JDK-8241248 code: 3010 ms

I can confirm that the synchronization code works. Without it I still catches exception from different threads trying to resume the same session from the cache

Reply:
Thank you for the update.

I would like to have the synchronization in the cache class for easier maintainence.  But please go ahead for the integration if you don't want to make the update now. We could file an enhancement later on.

Reply:
There were some changes that @ascarpino did for RSA blinding to reduce lock contention within that cache.  It doesn't look like we're running into a hot lock here based on the numbers above, but it might be worth looking at his solution down the road when we move synchronization into the Cache object.

Reply:
@XueleiFan , @jnimeh, thank you for your suggestions. I have updated synchronization using ReentrantLock in the Cache object. The performance test showed no issues.

Reply:
> @XueleiFan , @jnimeh, thank you for your suggestions. I have updated synchronization using ReentrantLock in the Cache object. The performance test showed no issues.

Thank you for considering use ReentrantLock, the "synchronized" keyword is something we are trying to avoid in the JSSE implementation.  However, the reentrant lock introduced is not the lock in the Cache implementation.  They are overlapped and could be conflict.  The performance test looks good with just a few threads, but I'm not very sure of the through put impact if the threads goes up to 1M or more (there were complaints previously for large volume traffics).

Maybe, you could keep it simple by adding a pull() method in the Cache class (find the session and remove it from the cache if found), by using the Cache class internal synchronize methods.

Reply:
Unfortunately, simple pull() can not be used in this case. We have to check if the session found in the cache can be rejoined with parameters received in ClientHello and server context. Only rejoinable sessions should be removed from the session cache.
It is possible to use simple pull() and restore session in the cache if the session is not rejoinable, but I do not like this approach. Also, it will require extending Cache with get/setExpirationTime methods.

Reply:
> Unfortunately, simple pull() can not be used in this case. We have to check if the session found in the cache can be rejoined with parameters received in ClientHello and server context. Only rejoinable sessions should be removed from the session cache.

For TLS 1.3, I think it may be safe to remove the session from the cache even if it is no rejoinable.

Reply:
@XueleiFan  Thank you for suggestion.  The patch is updated using a simple pull()